Support multiple attention groups for KV sharing

shinoss opened the pull request:
Summary:
https://github.com/vllm-project/vllm/pull/21588 added support for multiple attention metadata builders per kv-cache spec. 

As part of this change, each KV cache group now maps to one or more `AttentionGroup`, with one attention group being created for each type of attention backend used. 

However, if we want to enable KV sharing when we have more than one attention group, we run into the following assertion:
```
            assert len(attn_groups[group_idx]) == 1, (
                "Only one attention group per KV cache group is supported "
                "for KV-cache sharing for now.")
```

This PR adds support to make this implementation more flexible, such that we can support KV cache sharing when there are multiple attention groups per KV cache group.

Test Plan:
new added unit test passes:
```
pytest tests/v1/test_kv_sharing.py 
```

Rollback Plan:

Differential Revision: D80020191

heheda12345 commented on vllm/v1/worker/utils.py:
```diff
@@ -225,26 +225,34 @@ def initialize_kv_cache_for_kv_sharing(
-    for i, kv_cache_group in enumerate(kv_cache_groups):
-        for layer_name in kv_cache_group.layer_names:
-            layer_to_kv_cache_group_idx[layer_name] = i
+    layer_to_attn_group_idx: dict[str, int] = {}
+    if attn_groups:
+        for kv_cache_group_idx, kv_attn_groups in enumerate(attn_groups):
+            for attn_group_idx, attn_group in enumerate(kv_attn_groups):
+                for layer_name in attn_group.layer_names:
+                    layer_to_kv_cache_group_idx[
+                        layer_name] = kv_cache_group_idx
+                    layer_to_attn_group_idx[layer_name] = attn_group_idx
+    else:
+        for kv_cache_group_idx, kv_cache_group in enumerate(kv_cache_groups):
+            for layer_name in kv_cache_group.layer_names:
+                layer_to_kv_cache_group_idx[layer_name] = kv_cache_group_idx
```
only called once during startup so the performance isn't a big concern.

sarckk replied:
@Jialin multiple attn groups are currently used in cases where hybrid memory allocator is disabled but we have layers that require different attention backend (Llama4 for example, which has interleaved normal attention and chunked local attention layers). In that case we will have `len(self.attn_groups) == 1` and `self.attn_groups[0]` equal:

```
[
AttentionGroup(backend=FlashAttentionBackend, metadata_builder=FlashAttentionMetadataBuilder, layer_names=[all global layers]),
AttentionGroup(backend=ChunkedLocalFlashAttentionBackend, metadata_builder=ChunkedLocalFlashAttentionMetadataBuilder, layer_names=[all local layers])
]
```

hope that helps.

LucasWilkinson reviewed (approved):
LGTM; can you rebase off main to get a green CI

Having attn_groups as 2d array is getting a bit messy; we may want to re-think this but we can do that in a future PR if we decide to do something

heheda12345 reviewed (approved):
LGTM!

Commit:
Support multiple attention groups for KV sharing (#22672)

Summary:

https://github.com/vllm-project/vllm/pull/21588 added support for multiple attention metadata builders per kv-cache spec.

As part of this change, each KV cache group now maps to one or more `AttentionGroup`, with one attention group being created for each type of attention backend used.

However, if we want to enable KV sharing when we have more than one attention group, we run into the following assertion:
```
            assert len(attn_groups[group_idx]) == 1, (
                "Only one attention group per KV cache group is supported "
                "for KV-cache sharing for now.")
```

This PR adds support to make this implementation more flexible, such that we can support KV cache sharing when there are multiple attention groups per KV cache group.

Test Plan:
new added unit test passes:
```
pytest tests/v1/test_kv_sharing.py
```

Rollback Plan:

Differential Revision: D80020191

Signed-off-by: Yong Hoon Shin <yhshin@meta.com>
